test/functional/test_framework/wallet.py

```diff
@@ -71,9 +71,9 @@ def send_self_transfer(self, *, fee_rate=Decimal("0.003"), from_node, utxo_to_sp
         tx.wit.vtxinwit[0].scriptWitness.stack = [CScript([OP_TRUE])]
         tx_hex = tx.serialize().hex()
 
-        txid = from_node.sendrawtransaction(tx_hex)
-        self._utxos.append({'txid': txid, 'vout': 0, 'value': send_value})
-        tx_info = from_node.getmempoolentry(txid)
+        tx_info = from_node.testmempoolaccept([tx_hex])[0]
+        self._utxos.append({'txid': tx_info['txid'], 'vout': 0, 'value': send_value})
+        from_node.sendrawtransaction(tx_hex)
```

Comment:
I just have a conceptual question here. Why would we wanna send tx when we know it's going to fail? for instance I ran this code to check for failure and success and got
![markoPR2](https://user-images.githubusercontent.com/26356227/103995201-feda0f00-514c-11eb-87e5-180380da79d3.PNG). Is it to retrieve the error reason? If so, since `reject-reason` is a part of the `tx_info` object can't we just return the `reject-reason` rather than raising an exception?

Reply:
None of the code paths currently send a tx that we know is going to fail. Otherwise, `sendrawtransaction` would throw.

Reply:
This makes me notice a potentially unwanted side-effect in `send_self_transfer`. If a test case is catching an exception in `send_self_transfer`, `self._utxos.append({'txid': txid, 'vout': 0, 'value': send_value})` is happening before some rpc calls and assertions that could potentially fail. This could cause some confusing behavior in the case an exception does get thrown, and `self._utxos` is left modified

Not related to this PR imo as this behavior was already present, but do you think it would be useful to do `self._utxos.append({'txid': txid, 'vout': 0, 'value': send_value})` at the very end of this method, when we know it will succeed, in another PR?

Reply:
Well, I guess this would only apply if there are cases where `testmempoolaccept` succeeds but `sendrawtransaction` fails